aten/src/ATen/native/cuda/Embedding.cu

```diff
@@ -0,0 +1,348 @@
+  int64_t* indices, scalar_t* grad, scalar_t* grad_weight,
+  int64_t num_indices, int64_t stride, int padding_idx) {
+
+  const int feature_dim = blockIdx.x * 4 + threadIdx.x / 32;
+  if (feature_dim >= stride) {
+    return;
+  }
+
+  // The strategy here is that each warp handles a single feature
+  // dimension.
+  // Within that feature dimension, points in the [batch][element]
+  // dimension can overlap, and we need to determine if threads want
+  // to add to the gradient in a colliding manner.
+  // Typically one would use floating-point atomicAdd() to resolve
+  // these collisions, but that is non-deterministic if there are
+  // collisions. Non-determinism for this code is really bad,
+  // especially in RNNs, and is prone to snowballing error.
+  // In order to get a deterministic order of execution, we handle
+  // non-colliding updates separately from colliding ones. Colliding
+  // updates are serialized in their order of execution by using the
+  // warp-wide collision detector `warp_has_collision`.
+  const int laneId = threadIdx.x % 32;
+  for (int64_t i = laneId; i < num_indices; i += WARP_SIZE) {
+    const int weight_index = (int)indices[i];
+    if (weight_index == padding_idx) {
+      continue;
+    }
+
+    auto value = grad[i * stride + feature_dim];
+
+    // FIXME: should we accumulate as accreal?
+    // Check for collision
+    if (warp_has_collision(weight_index)) {
+      // Run all lanes sequentially; warp divergence
+      for (int i = 0; i < WARP_SIZE; ++i) {
+        if (laneId == i) {
+          grad_weight[weight_index * stride + feature_dim] += value;
+        }
+      }
+    } else {
+      // No collision; warp coherence
+      grad_weight[weight_index * stride + feature_dim] += value;
+    }
+  }
+}
+
+
+template <typename scalar_t>
+__global__ void embedding_backward_kernel(
+  int64_t* input, int64_t* indices, scalar_t* grad_output, scalar_t* grad_weight,
+  int64_t* count, int64_t numel, int64_t stride, int padding_idx) {
+
+  using accscalar_t = cuda::acc_type<scalar_t>;
+  int idx = blockIdx.x * 4 + threadIdx.y;
+
+  // Each warp is responsible for an input into the LookupTable.
+  // If the preceding input has the same as this input, then the warp
+  // exits immediately. The warp also processes subsequent inputs with the
+  // same value.
+  //
+  // Input Warp
+  // 1     <warp 1>
+  // 1     <warp 1> (<warp 2> exits without doing any work)
+  // 5     <warp 3>
+  // 8     <warp 4>
+
+  // Number of values proceessed by each thread (grain size)
+  const int SZ = 4;
+
+  if (idx < numel
+      && (idx == 0 || input[idx] != input[idx - 1])
+      && input[idx] != padding_idx) {
+    do {
+      const int start_feature = threadIdx.x + blockIdx.y * blockDim.x * SZ;
+      const int weight_row = ((int) input[idx]) * stride;
+      const int grad_row = ((int) indices[idx]) * stride;
+      const accscalar_t scale = count ? (accscalar_t)1.0 / count[idx] : 1.0;
```

Comment:
why do you need the conversion on only one of the cases? (there's no warning on the false case?)